flink-core/src/main/java/org/apache/flink/core/memory/HybridMemorySegment.java

```diff
@@ -94,26 +126,22 @@
     @Override
     public void free() {
         super.free();
+        if (cleaner != null) {
+            cleaner.run();
+        }
```

Comment:
Looking at the special casing of this class, why don't we introduce an `DirectMemorySegment` and a `UnsafeMemorySegment`? Also, why do we have the `HeapMemorySegment` and still allow the `HybridMemorySegment` to be used with heap memory?

Reply:
If we had a `DirectMemorySegment`, then we could only allow `wrap` on this type, for example.

Reply:
`HybridMemorySegment` is currently the only `MemorySegment` implementation that is used in production. `HeapMemorySegment` is no longer used, but seems intentionally not removed as it is annotated with `@SuppressWarnings("unused")`.

From my side, it makes sense to split `HybridMemorySegment` into `HeapMemorySegment`, `DirectMemorySegment` and `UnsafeMemorySegment`. This should get rid of lots of `if-else` branches in the implementation.

I'm leaning towards addressing this refactor as a follow-up effort, and only do that for the master branch.

Reply:
Ok, this sounds reasonable. Then let's do this refactoring soon. In fact, `HeapMemorySegment` does not seem to be that unused anymore. Somehow the table API started using this class for some `RawFormatDeserializationSchema`. This might be a good indicator for pulling things straight.

Reply:
With this refactoring we could also think about getting rid of the whole native memory cleaner logic.

Reply:
I'll drive the refactor effort and try to get it done in the 1.13 release cycle.